File: backend/app/api/people.py
"""API endpoints for people intelligence."""

from uuid import UUID
from fastapi import APIRouter, HTTPException, status, Query
from pydantic import BaseModel

from app.api.deps import Database, CurrentUser
from app.services.people_service import PeopleService


router = APIRouter()


class PersonSummary(BaseModel):
    """Summary of a person in the list."""
    id: str
    name: str
    entity_type: str
    email: str | None
    mention_count: int
    first_seen: str | None
    last_seen: str | None


class ContactSuggestion(BaseModel):
    """Contact for autocomplete."""
    id: str
    name: str
    email: str | None
    mention_count: int


class ContactSearchResponse(BaseModel):
    """Response for contact search/autocomplete."""
    contacts: list[ContactSuggestion]


class PeopleListResponse(BaseModel):
    """Response for listing people."""
    people: list[PersonSummary]
    total: int


class MemoryBrief(BaseModel):
    """Brief memory reference."""
    id: str
    content: str
    memory_type: str
    memory_date: str | None


class PersonProfileResponse(BaseModel):
    """Full person profile response."""
    name: str
    entity_type: str
    email: str | None
    mention_count: int
    first_seen: str | None
    last_seen: str | None
    summary: str | None
    relationship_type: str | None
    topics: list[str]
    sentiment_trend: str | None
    last_interaction_date: str | None
    next_meeting_date: str | None
    recent_memories: list[MemoryBrief]


class PersonMemoriesResponse(BaseModel):
    """Response for person's memories."""
    memories: list[MemoryBrief]
    total: int


class MeetingContextResponse(BaseModel):
    """Meeting context response."""
    person_name: str
    context: str | None


@router.get("/search", response_model=ContactSearchResponse)
async def search_contacts(
    current_user: CurrentUser,
    db: Database,
    q: str = Query("", description="Search query for name or email"),
    limit: int = Query(10, ge=1, le=20),
):
    """
    Search for contacts by name or email for autocomplete.

    Returns contacts with email addresses, sorted by relevance.
    Use this for email composition to suggest recipients.
    """
    people_service = PeopleService(db)
    contacts = await people_service.search_contacts(
        user_id=current_user.id,
        query=q,
        limit=limit,
    )

    return ContactSearchResponse(
        contacts=[ContactSuggestion(**c) for c in contacts]
    )


@router.get("", response_model=PeopleListResponse)
async def list_people(
    current_user: CurrentUser,
    db: Database,
    sort_by: str = Query("recent", enum=["recent", "frequent", "alphabetical"]),
    limit: int = Query(50, ge=1, le=100),
):
    """
    List all people the user knows about.

    Sort options:
    - recent: Most recently mentioned first
    - frequent: Most frequently mentioned first
    - alphabetical: Alphabetically by name
    """
    people_service = PeopleService(db)
    people = await people_service.list_people(
        user_id=current_user.id,
        sort_by=sort_by,
        limit=limit,
    )

    return PeopleListResponse(
        people=[PersonSummary(**p) for p in people],
        total=len(people),
    )


@router.get("/{person_name}", response_model=PersonProfileResponse)
async def get_person_profile(
    person_name: str,
    current_user: CurrentUser,
    db: Database,
    regenerate: bool = Query(False, description="Force regenerate the profile"),
):
    """
    Get comprehensive profile for a person.

    The profile includes:
    - Basic stats (mention count, first/last seen)
    - AI-generated summary of the relationship
    - Common topics discussed
    - Sentiment trend
    - Recent memories
    """
    people_service = PeopleService(db)
    profile = await people_service.get_person_profile(
        user_id=current_user.id,
        person_name=person_name,
        regenerate=regenerate,
    )

    if not profile:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Person '{person_name}' not found",
        )

    return PersonProfileResponse(**profile)


@router.get("/{person_name}/memories", response_model=PersonMemoriesResponse)
async def get_person_memories(
    person_name: str,
    current_user: CurrentUser,
    db: Database,
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    """
    Get all memories mentioning a person.
    """
    people_service = PeopleService(db)
    memories = await people_service.get_person_memories(
        user_id=current_user.id,
        person_name=person_name,
        limit=limit,
        offset=offset,
    )

    return PersonMemoriesResponse(
        memories=[
            MemoryBrief(
                id=str(m.id),
                content=m.content[:200],
                memory_type=m.memory_type,
                memory_date=m.memory_date.isoformat() if m.memory_date else None,
            )
            for m in memories
        ],
        total=len(memories),
    )


@router.get("/{person_name}/context", response_model=MeetingContextResponse)
async def get_meeting_context(
    person_name: str,
    current_user: CurrentUser,
    db: Database,
):
    """
    Get meeting preparation context for a person.

    Returns bullet points summarizing:
    - Key topics discussed recently
    - Pending items or follow-ups
    - Useful context for the meeting
    """
    people_service = PeopleService(db)

    # First verify person exists
    entity = await people_service.get_person_by_name(current_user.id, person_name)
    if not entity:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Person '{person_name}' not found",
        )

    context = await people_service.generate_meeting_context(
        user_id=current_user.id,
        person_name=person_name,
    )

    return MeetingContextResponse(
        person_name=person_name,
        context=context,
    )

File: mobile/src/services/people.ts
/**
 * People Service
 *
 * @deprecated This service is deprecated. Use entities.ts instead.
 *
 * This file now wraps the entities service for backwards compatibility.
 * The backend has replaced /people endpoints with /v3/entities.
 */

import { entitiesService, Entity, EntityMemoriesResponse } from './entities';
import { logger } from '../utils/logger';

// Legacy types for backwards compatibility
export interface ContactSuggestion {
  id: string;
  name: string;
  email: string | null;
  mention_count: number;
}

export interface ContactSearchResponse {
  contacts: ContactSuggestion[];
}

export interface PersonSummary {
  id: string;
  name: string;
  entity_type: string;
  email: string | null;
  mention_count: number;
  first_seen: string | null;
  last_seen: string | null;
}

export interface PersonProfile {
  id: string;
  name: string;
  entity_type: string;
  email: string | null;
  mention_count: number;
  first_seen: string | null;
  last_seen: string | null;
  summary: string | null;
  relationship_type: string | null;
  topics: string[];
  sentiment_trend: string | null;
  last_interaction_date: string | null;
  next_meeting_date: string | null;
  recent_memories: Array<{
    id: string;
    content: string;
    memory_type: string;
    memory_date: string | null;
  }>;
}

export interface PeopleListResponse {
  people: PersonSummary[];
  total: number;
}

export interface PersonMemoriesResponse {
  memories: Array<{
    id: string;
    content: string;
    created_at: string;
  }>;
  total: number;
}

// Transform Entity to PersonSummary
function entityToPersonSummary(entity: Entity): PersonSummary {
  return {
    id: entity.id,
    name: entity.name,
    entity_type: entity.entity_type,
    email: entity.email || null,
    mention_count: entity.mention_count,
    first_seen: entity.first_seen,
    last_seen: entity.last_seen,
  };
}

// Transform Entity to PersonProfile
function entityToPersonProfile(entity: Entity, memories: Array<{ id: string; content: string; created_at: string }> = []): PersonProfile {
  const attrs = entity.attributes as Record<string, any> || {};
  return {
    id: entity.id,
    name: entity.name,
    entity_type: entity.entity_type,
    email: entity.email || null,
    mention_count: entity.mention_count,
    first_seen: entity.first_seen,
    last_seen: entity.last_seen,
    summary: attrs.summary || `${entity.name} is mentioned in ${entity.mention_count} memories.`,
    relationship_type: attrs.relationship_type || attrs.relationship || null,
    topics: attrs.topics || attrs.recent_topics || [],
    sentiment_trend: attrs.sentiment_trend || null,
    last_interaction_date: entity.last_seen,
    next_meeting_date: attrs.next_meeting_date || null,
    recent_memories: memories.map(m => ({
      id: m.id,
      content: m.content,
      memory_type: 'text',
      memory_date: m.created_at,
    })),
  };
}

class PeopleService {
  /**
   * @deprecated Use entitiesService.getEntities({ type: 'person' }) instead
   */
  async listPeople(
    _sortBy: 'recent' | 'frequent' | 'alphabetical' = 'recent',
    limit: number = 50
  ): Promise<PeopleListResponse> {
    logger.warn('PeopleService: listPeople is deprecated, use entitiesService.getEntities()');

    const response = await entitiesService.getEntities({
      type: 'person',
      limit,
    });

    return {
      people: response.entities.map(entityToPersonSummary),
      total: response.total,
    };
  }

  /**
   * @deprecated Use entitiesService.getEntity() instead
   */
  async getPersonProfile(
    name: string,
    _regenerate: boolean = false
  ): Promise<PersonProfile> {
    logger.warn('PeopleService: getPersonProfile is deprecated, use entitiesService');

    // Search for the person by name
    const entity = await entitiesService.getPersonByName(name);
    if (!entity) {
      throw new Error(`Person "${name}" not found`);
    }

    // Also fetch recent memories for this person
    let memories: Array<{ id: string; content: string; created_at: string }> = [];
    try {
      const memoriesResponse = await entitiesService.getEntityMemories(entity.id, { limit: 5 });
      memories = memoriesResponse.memories.map(m => ({
        id: m.id,
        content: m.content,
        created_at: m.created_at,
      }));
    } catch (err) {
      logger.warn('Failed to fetch memories for person:', err);
    }

    return entityToPersonProfile(entity, memories);
  }

  /**
   * @deprecated Use entitiesService.getEntityMemories() instead
   */
  async getPersonMemories(
    name: string,
    limit: number = 20,
    offset: number = 0
  ): Promise<PersonMemoriesResponse> {
    logger.warn('PeopleService: getPersonMemories is deprecated, use entitiesService');

    // Search for the person by name
    const entity = await entitiesService.getPersonByName(name);
    if (!entity) {
      return { memories: [], total: 0 };
    }

    const response = await entitiesService.getEntityMemories(entity.id, { limit, offset });
    return {
      memories: response.memories.map(m => ({
        id: m.id,
        content: m.content,
        created_at: m.created_at,
      })),
      total: response.total,
    };
  }

  /**
   * @deprecated Meeting context not available in v3 API
   */
  async getMeetingContext(name: string): Promise<{
    context: string | null;
    topics: string[];
    last_meeting: string | null;
  }> {
    logger.warn('PeopleService: getMeetingContext is deprecated');

    // Try to get entity info and recent memories
    const entity = await entitiesService.getPersonByName(name);
    if (!entity) {
      return {
        context: `No context available for ${name}`,
        topics: [],
        last_meeting: null,
      };
    }

    // Fetch recent memories to build context
    let contextParts: string[] = [];
    try {
      const memoriesResponse = await entitiesService.getEntityMemories(entity.id, { limit: 5 });
      contextParts = memoriesResponse.memories.slice(0, 3).map(m => m.content.slice(0, 200));
    } catch (err) {
      logger.warn('Failed to fetch memories for meeting context:', err);
    }

    const attrs = entity.attributes as Record<string, any> || {};
    const topics = attrs.topics || [];

    return {
      context: contextParts.length > 0
        ? `Recent interactions with ${entity.name}:\n${contextParts.join('\n\n')}`
        : `${entity.name} - mentioned ${entity.mention_count} times`,
      topics,
      last_meeting: entity.last_seen,
    };
  }

  /**
   * @deprecated Connections not available in v3 API
   */
  async listConnections(
    _limit: number = 20,
    _unnotifiedOnly: boolean = false
  ): Promise<{ connections: any[]; total: number }> {
    logger.warn('PeopleService: listConnections is deprecated and not available');
    return { connections: [], total: 0 };
  }

  /**
   * @deprecated Connections not available in v3 API
   */
  async getConnection(_connectionId: string): Promise<null> {
    logger.warn('PeopleService: getConnection is deprecated and not available');
    return null;
  }

  /**
   * @deprecated Connections not available in v3 API
   */
  async dismissConnection(_connectionId: string): Promise<{ success: boolean }> {
    logger.warn('PeopleService: dismissConnection is deprecated and not available');
    return { success: false };
  }

  /**
   * @deprecated Use entitiesService.searchEntities() instead
   */
  async searchContacts(
    query: string = '',
    limit: number = 10
  ): Promise<ContactSearchResponse> {
    logger.warn('PeopleService: searchContacts is deprecated, use entitiesService.searchEntities()');

    const response = await entitiesService.searchContacts(query, limit);
    return {
      contacts: response.contacts.map(entity => ({
        id: entity.id,
        name: entity.name,
        email: (entity.attributes as any)?.email || null,
        mention_count: entity.mention_count,
      })),
    };
  }
}

export const peopleService = new PeopleService();
